File: LudoGame/LudoGame/Engine/LudoGameEngine.swift
import Foundation

struct LudoGameEngine {
    private(set) var state: LudoGameState

    init(playerCount: Int = 4) {
        state = LudoGameState(playerCount: min(max(playerCount, 2), 4))
    }

    mutating func rollDice() {
        guard state.winner == nil, !state.isRolling else { return }
        guard state.selectableTokenIDs.isEmpty else { return }

        state.isRolling = true
        let roll = Int.random(in: 1...6)
        state.lastDiceRoll = roll

        if roll == 6 {
            state.consecutiveSixes += 1
        } else {
            state.consecutiveSixes = 0
        }

        if state.consecutiveSixes >= 3 {
            state.statusMessage = "\(state.currentPlayer.id.displayName) rolled three 6s — turn skipped"
            state.consecutiveSixes = 0
            state.isRolling = false
            advanceTurn(extraTurn: false)
            return
        }

        let movable = movableTokens(for: state.currentPlayer.id, dice: roll)
        if movable.isEmpty {
            state.statusMessage = "\(state.currentPlayer.id.displayName) rolled \(roll) — no valid moves"
            state.isRolling = false
            advanceTurn(extraTurn: roll == 6)
            return
        }

        if movable.count == 1, let only = movable.first {
            state.statusMessage = "\(state.currentPlayer.id.displayName) rolled \(roll)"
            moveToken(only, dice: roll)
            return
        }

        state.selectableTokenIDs = Set(movable)
        state.statusMessage = "\(state.currentPlayer.id.displayName) rolled \(roll) — choose a token"
        state.isRolling = false
    }

    mutating func selectToken(_ tokenID: Int) {
        guard state.selectableTokenIDs.contains(tokenID),
              let dice = state.lastDiceRoll else { return }

        state.selectableTokenIDs.removeAll()
        moveToken(tokenID, dice: dice)
    }

    private mutating func moveToken(_ tokenID: Int, dice: Int) {
        guard let playerIndex = state.players.firstIndex(where: { $0.id == state.currentPlayer.id }),
              let tokenIndex = state.players[playerIndex].tokens.firstIndex(where: { $0.id == tokenID }) else {
            return
        }

        var token = state.players[playerIndex].tokens[tokenIndex]
        let previousAbsolute = LudoBoardLayout.absoluteTrackIndex(for: token)
        let captured = applyMove(to: &token, dice: dice)
        state.players[playerIndex].tokens[tokenIndex] = token

        if state.players[playerIndex].hasWon {
            state.winner = state.players[playerIndex].id
            state.statusMessage = "\(state.winner!.displayName) wins!"
            state.isRolling = false
            return
        }

        let extraTurn = dice == 6 || captured
        if captured {
            state.statusMessage = "\(state.currentPlayer.id.displayName) captured a token!"
        } else if token.isFinished {
            state.statusMessage = "\(state.currentPlayer.id.displayName) reached home!"
        } else {
            state.statusMessage = "\(state.currentPlayer.id.displayName) moved token \(tokenID + 1)"
        }

        _ = previousAbsolute
        state.isRolling = false
        advanceTurn(extraTurn: extraTurn)
    }

    private mutating func applyMove(to token: inout GameToken, dice: Int) -> Bool {
        if token.isInYard {
            guard dice == 6 else { return false }
            token.position = 0
            return captureIfNeeded(at: token)
        }

        if token.isOnMainTrack {
            let newPosition = token.position + dice
            if newPosition > LudoBoardLayout.pathLength - 1 {
                let overflow = newPosition - (LudoBoardLayout.pathLength - 1)
                if overflow > LudoBoardLayout.homeStretchLength {
                    return false
                }
                token.position = LudoBoardLayout.pathLength - 1 + overflow
                return false
            }
            token.position = newPosition
            return captureIfNeeded(at: token)
        }

        if token.isInHomeStretch {
            let homeIndex = token.position - LudoBoardLayout.pathLength + 1
            let newHomeIndex = homeIndex + dice
            if newHomeIndex > LudoBoardLayout.homeStretchLength {
                return false
            }
            if newHomeIndex == LudoBoardLayout.homeStretchLength {
                token.position = 58
            } else {
                token.position = LudoBoardLayout.pathLength - 1 + newHomeIndex
            }
            return false
        }

        return false
    }

    private mutating func captureIfNeeded(at token: GameToken) -> Bool {
        guard token.isOnMainTrack,
              let absolute = LudoBoardLayout.absoluteTrackIndex(for: token),
              !PlayerColor.safePositions.contains(absolute) else {
            return false
        }

        var captured = false
        for playerIndex in state.players.indices {
            guard state.players[playerIndex].id != token.owner else { continue }
            for tokenIndex in state.players[playerIndex].tokens.indices {
                var opponent = state.players[playerIndex].tokens[tokenIndex]
                guard opponent.isOnMainTrack,
                      let opponentAbsolute = LudoBoardLayout.absoluteTrackIndex(for: opponent),
                      opponentAbsolute == absolute else { continue }
                opponent.position = -1
                state.players[playerIndex].tokens[tokenIndex] = opponent
                captured = true
            }
        }
        return captured
    }

    private mutating func advanceTurn(extraTurn: Bool) {
        state.lastDiceRoll = nil
        state.selectableTokenIDs.removeAll()

        guard state.winner == nil else { return }
        guard !extraTurn else {
            state.statusMessage += " — roll again!"
            return
        }

        repeat {
            state.currentPlayerIndex = (state.currentPlayerIndex + 1) % state.players.count
        } while !state.players[state.currentPlayerIndex].isActive || state.players[state.currentPlayerIndex].hasWon
    }

    private func movableTokens(for player: PlayerColor, dice: Int) -> [Int] {
        guard let gamePlayer = state.players.first(where: { $0.id == player }) else { return [] }
        return gamePlayer.tokens.compactMap { token in
            canMove(token, dice: dice) ? token.id : nil
        }
    }

    private func canMove(_ token: GameToken, dice: Int) -> Bool {
        if token.isFinished { return false }
        if token.isInYard { return dice == 6 }

        if token.isOnMainTrack {
            let newPosition = token.position + dice
            if newPosition > LudoBoardLayout.pathLength - 1 {
                let overflow = newPosition - (LudoBoardLayout.pathLength - 1)
                return overflow <= LudoBoardLayout.homeStretchLength
            }
            return true
        }

        if token.isInHomeStretch {
            let homeIndex = token.position - LudoBoardLayout.pathLength + 1
            return homeIndex + dice <= LudoBoardLayout.homeStretchLength
        }

        return false
    }
}

File: LudoGameAndroid/app/src/main/java/com/ludogame/app/engine/LudoGameEngine.kt
package com.ludogame.app.engine

import com.ludogame.app.models.GameToken
import com.ludogame.app.models.LudoGameState
import com.ludogame.app.models.PlayerColor
import kotlin.random.Random

class LudoGameEngine(playerCount: Int = 4) {
    var state: LudoGameState = LudoGameState(playerCount)
        private set

    fun rollDice() {
        val current = state
        if (current.winner != null || current.isRolling) return
        if (current.selectableTokenIDs.isNotEmpty()) return

        state = current.copy(isRolling = true)
        val roll = Random.nextInt(1, 7)
        val consecutiveSixes = if (roll == 6) current.consecutiveSixes + 1 else 0

        if (consecutiveSixes >= 3) {
            state = state.copy(
                consecutiveSixes = 0,
                isRolling = false,
                statusMessage = "${current.currentPlayer.id.displayName} rolled three 6s — turn skipped"
            )
            advanceTurn(extraTurn = false)
            return
        }

        val movable = movableTokens(current.currentPlayer.id, roll)
        if (movable.isEmpty()) {
            state = state.copy(
                lastDiceRoll = roll,
                consecutiveSixes = consecutiveSixes,
                isRolling = false,
                statusMessage = "${current.currentPlayer.id.displayName} rolled $roll — no valid moves"
            )
            advanceTurn(extraTurn = roll == 6)
            return
        }

        if (movable.size == 1) {
            state = state.copy(
                lastDiceRoll = roll,
                consecutiveSixes = consecutiveSixes,
                statusMessage = "${current.currentPlayer.id.displayName} rolled $roll"
            )
            moveToken(movable.first(), roll)
            return
        }

        state = state.copy(
            lastDiceRoll = roll,
            consecutiveSixes = consecutiveSixes,
            selectableTokenIDs = movable.toSet(),
            statusMessage = "${current.currentPlayer.id.displayName} rolled $roll — choose a token",
            isRolling = false
        )
    }

    fun selectToken(tokenId: Int) {
        val current = state
        if (!current.selectableTokenIDs.contains(tokenId)) return
        val dice = current.lastDiceRoll ?: return

        state = current.copy(selectableTokenIDs = emptySet())
        moveToken(tokenId, dice)
    }

    private fun moveToken(tokenId: Int, dice: Int) {
        val current = state
        val playerIndex = current.players.indexOfFirst { it.id == current.currentPlayer.id }
        if (playerIndex < 0) return

        val player = current.players[playerIndex]
        val tokenIndex = player.tokens.indexOfFirst { it.id == tokenId }
        if (tokenIndex < 0) return

        val token = player.tokens[tokenIndex]
        val (updatedToken, captured) = applyMove(token, dice)
        val updatedTokens = player.tokens.toMutableList().apply { this[tokenIndex] = updatedToken }
        val updatedPlayer = player.copy(tokens = updatedTokens)
        val updatedPlayers = current.players.toMutableList().apply { this[playerIndex] = updatedPlayer }

        if (updatedPlayer.hasWon) {
            state = current.copy(
                players = updatedPlayers,
                winner = updatedPlayer.id,
                statusMessage = "${updatedPlayer.id.displayName} wins!",
                isRolling = false
            )
            return
        }

        val extraTurn = dice == 6 || captured
        val statusMessage = when {
            captured -> "${current.currentPlayer.id.displayName} captured a token!"
            updatedToken.isFinished -> "${current.currentPlayer.id.displayName} reached home!"
            else -> "${current.currentPlayer.id.displayName} moved token ${tokenId + 1}"
        }

        state = current.copy(
            players = updatedPlayers,
            statusMessage = statusMessage,
            isRolling = false
        )
        advanceTurn(extraTurn = extraTurn)
    }

    private fun applyMove(token: GameToken, dice: Int): Pair<GameToken, Boolean> {
        return when {
            token.isInYard -> {
                if (dice != 6) return token to false
                val moved = token.copy(position = 0)
                moved to captureIfNeeded(moved)
            }
            token.isOnMainTrack -> {
                val newPosition = token.position + dice
                if (newPosition > LudoBoardLayout.PATH_LENGTH - 1) {
                    val overflow = newPosition - (LudoBoardLayout.PATH_LENGTH - 1)
                    if (overflow > LudoBoardLayout.HOME_STRETCH_LENGTH) {
                        return token to false
                    }
                    val moved = token.copy(position = LudoBoardLayout.PATH_LENGTH - 1 + overflow)
                    return moved to false
                }
                val moved = token.copy(position = newPosition)
                moved to captureIfNeeded(moved)
            }
            token.isInHomeStretch -> {
                val homeIndex = token.position - LudoBoardLayout.PATH_LENGTH + 1
                val newHomeIndex = homeIndex + dice
                if (newHomeIndex > LudoBoardLayout.HOME_STRETCH_LENGTH) {
                    return token to false
                }
                val newPosition = if (newHomeIndex == LudoBoardLayout.HOME_STRETCH_LENGTH) {
                    58
                } else {
                    LudoBoardLayout.PATH_LENGTH - 1 + newHomeIndex
                }
                token.copy(position = newPosition) to false
            }
            else -> token to false
        }
    }

    private fun captureIfNeeded(token: GameToken): Boolean {
        if (!token.isOnMainTrack) return false
        val absolute = LudoBoardLayout.absoluteTrackIndex(token) ?: return false
        if (PlayerColor.safePositions.contains(absolute)) return false

        var captured = false
        val updatedPlayers = state.players.map { player ->
            if (player.id == token.owner) return@map player

            val updatedTokens = player.tokens.map { opponent ->
                if (!opponent.isOnMainTrack) return@map opponent
                val opponentAbsolute = LudoBoardLayout.absoluteTrackIndex(opponent) ?: return@map opponent
                if (opponentAbsolute == absolute) {
                    captured = true
                    opponent.copy(position = -1)
                } else {
                    opponent
                }
            }
            player.copy(tokens = updatedTokens)
        }

        if (captured) {
            state = state.copy(players = updatedPlayers)
        }
        return captured
    }

    private fun advanceTurn(extraTurn: Boolean) {
        val current = state
        state = current.copy(
            lastDiceRoll = null,
            selectableTokenIDs = emptySet()
        )

        if (state.winner != null) return
        if (extraTurn) {
            state = state.copy(statusMessage = "${state.statusMessage} — roll again!")
            return
        }

        var nextIndex = state.currentPlayerIndex
        do {
            nextIndex = (nextIndex + 1) % state.players.size
        } while (!state.players[nextIndex].isActive || state.players[nextIndex].hasWon)

        state = state.copy(currentPlayerIndex = nextIndex)
    }

    private fun movableTokens(player: PlayerColor, dice: Int): List<Int> {
        val gamePlayer = state.players.firstOrNull { it.id == player } ?: return emptyList()
        return gamePlayer.tokens.filter { canMove(it, dice) }.map { it.id }
    }

    private fun canMove(token: GameToken, dice: Int): Boolean {
        if (token.isFinished) return false
        if (token.isInYard) return dice == 6

        if (token.isOnMainTrack) {
            val newPosition = token.position + dice
            if (newPosition > LudoBoardLayout.PATH_LENGTH - 1) {
                val overflow = newPosition - (LudoBoardLayout.PATH_LENGTH - 1)
                return overflow <= LudoBoardLayout.HOME_STRETCH_LENGTH
            }
            return true
        }

        if (token.isInHomeStretch) {
            val homeIndex = token.position - LudoBoardLayout.PATH_LENGTH + 1
            return homeIndex + dice <= LudoBoardLayout.HOME_STRETCH_LENGTH
        }

        return false
    }

    fun restart(playerCount: Int) {
        state = LudoGameState(playerCount.coerceIn(2, 4))
    }
}
